File: articles/metabolism-energy-biological-function/c/metabolism_kernel.c
/*
 * Compact metabolism kernel in C.
 */

#include <math.h>
#include <stdio.h>

double exponential_growth(double t, double n0, double r) {
    return n0 * exp(r * t);
}

double logistic_growth(double t, double n0, double r, double k) {
    return k / (1.0 + ((k - n0) / n0) * exp(-r * t));
}

double doubling_time(double r) {
    if (r <= 0.0) {
        return NAN;
    }
    return log(2.0) / r;
}

double monod_growth(double substrate, double mu_max, double ks) {
    return mu_max * substrate / (ks + substrate);
}

double biomass_yield(double delta_x, double delta_s) {
    return delta_x / delta_s;
}

double toy_objective(double biomass_flux, double product_flux) {
    return biomass_flux + 0.25 * product_flux;
}

int main(void) {
    double n0 = 1.0e5;
    double r = log(4.0) / 48.0;

    printf("abundance_48h=%.6f\n", exponential_growth(48.0, n0, r));
    printf("doubling_time_h=%.6f\n", doubling_time(r));
    printf("logistic_96h=%.6f\n", logistic_growth(96.0, n0, 0.035, 1.0e6));
    printf("monod_growth=%.6f\n", monod_growth(5.0, 0.08, 2.5));
    printf("biomass_yield=%.6f\n", biomass_yield(0.75, 1.50));
    printf("maintenance_fraction=%.6f\n", 0.70 / 2.0);
    printf("toy_objective=%.6f\n", toy_objective(8.0, 2.0));

    return 0;
}

File: articles/metabolism-energy-biological-function/fortran/metabolism_kernel.f90
! Compact metabolism numerical kernel in Fortran.

program metabolism_kernel
  implicit none

  real :: n0
  real :: n48
  real :: r
  real :: td
  real :: K
  real :: logistic_96
  real :: substrate
  real :: mu_max
  real :: Ks
  real :: mu
  real :: delta_x
  real :: delta_s
  real :: Yxs
  real :: maintenance
  real :: substrate_input
  real :: maintenance_fraction
  real :: biomass_flux
  real :: product_flux
  real :: objective

  n0 = 1.0e5
  r = log(4.0) / 48.0
  n48 = n0 * exp(r * 48.0)
  td = log(2.0) / r

  K = 1.0e6
  logistic_96 = K / (1.0 + ((K - n0) / n0) * exp(-0.035 * 96.0))

  substrate = 5.0
  mu_max = 0.08
  Ks = 2.5
  mu = mu_max * substrate / (Ks + substrate)

  delta_x = 0.75
  delta_s = 1.50
  Yxs = delta_x / delta_s

  maintenance = 0.70
  substrate_input = 2.0
  maintenance_fraction = maintenance / substrate_input

  biomass_flux = 8.0
  product_flux = 2.0
  objective = biomass_flux + 0.25 * product_flux

  print *, "Exponential abundance at 48h:", n48
  print *, "Doubling time:", td
  print *, "Logistic abundance at 96h:", logistic_96
  print *, "Monod growth rate:", mu
  print *, "Biomass yield:", Yxs
  print *, "Maintenance fraction:", maintenance_fraction
  print *, "Toy flux objective:", objective
end program metabolism_kernel
